cni/ipam/ipam_test.go

```diff
@@ -27,70 +30,192 @@ var ipamQueryResponse = "" +
+
+			It("Start IPAM plugin", func() {
+				// Configure test mode.
+				plugin.SetOption(common.OptEnvironment, common.OptEnvironmentAzure)
+				plugin.SetOption(common.OptAPIServerURL, "null")
+				plugin.SetOption(common.OptIpamQueryUrl, "http://"+ipamQueryUrl)
+				// Start the plugin.
+				err = plugin.Start(&config)
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
+
+		Describe("Test IPAM ADD and DELETE pool", func() {
+
+			var result *cniTypesCurr.Result
+
+			Context("When ADD with nothing (request pool)", func() {
+				It("Request pool and ADD successfully", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "", "")
+					err = plugin.Add(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+					result, err = parseResult(arg.StdinData)
+					Expect(err).ShouldNot(HaveOccurred())
+					address1, _ := platform.ConvertStringToIPNet("10.0.0.5/16")
+					address2, _ := platform.ConvertStringToIPNet("10.0.0.6/16")
+					Expect(result.IPs[0].Address.IP).Should(Or(Equal(address1.IP), Equal(address2.IP)))
+					Expect(result.IPs[0].Address.Mask).Should(Equal(address1.Mask))
+				})
+			})
+
+			Context("When DELETE with subnet and address", func() {
+				It("DELETE address successfully", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "10.0.0.0/16", result.IPs[0].Address.IP.String())
+					err = plugin.Delete(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+			})
+
+			Context("When DELETE with subnet (release pool)", func() {
+				It("DELETE pool successfully", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "10.0.0.0/16", "")
+					err = plugin.Delete(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+			})
+		})
+
+		Describe("Test IPAM ADD and DELETE pool", func() {
+
+			Context("When address and subnet is given", func() {
+				It("ADD address successfully with the given address", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "10.0.0.0/16", "10.0.0.6")
+					err = plugin.Add(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+					result, err := parseResult(arg.StdinData)
+					Expect(err).ShouldNot(HaveOccurred())
+					address, _ := platform.ConvertStringToIPNet("10.0.0.6/16")
+					Expect(result.IPs[0].Address.IP).Should(Equal(address.IP))
+					Expect(result.IPs[0].Address.Mask).Should(Equal(address.Mask))
+				})
+			})
+
+			Context("When subnet is given", func() {
+				It("ADD successfully with a usable address", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "10.0.0.0/16", "")
+					err = plugin.Add(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+					result, err := parseResult(arg.StdinData)
+					Expect(err).ShouldNot(HaveOccurred())
+					address, _ := platform.ConvertStringToIPNet("10.0.0.5/16")
+					Expect(result.IPs[0].Address.IP).Should(Equal(address.IP))
+					Expect(result.IPs[0].Address.Mask).Should(Equal(address.Mask))
+				})
+			})
+		})
+
+		Describe("Test IPAM DELETE", func() {
+
+			Context("When address and subnet is given", func() {
```

Comment:
We may need tests to cover specifying an address and empty string.

Reply:
I'll add a case of release pool, when address is empty

Reply:
also add a case where release pool fails when address is allocated in pool and not released